command/ca/policy/policycontext/context.go

```diff
@@ -0,0 +1,186 @@
+func isPolicyLevel(ctx context.Context, level policyLevel) bool {
+	if ctx == nil {
+		return false
+	}
+	v, _ := ctx.Value(policyLevelContextKey{}).(policyLevel)
+	return v == level
+}
+
+type policyConfigurationTypeContextKey struct{}
+
+type policyConfigurationType int
+
+const (
+	x509PolicyType policyConfigurationType = iota + 1
+	sshHostPolicyType
+	sshUserPolicyType
+)
+
+// WithX509Policy returns a new context.Context with
+// parent ctx and X509 policy set.
+func WithX509Policy(ctx context.Context) context.Context {
+	return context.WithValue(ctx, policyConfigurationTypeContextKey{}, x509PolicyType)
+}
+
+// IsX509Policy returns if the context.Context has X.509 policy set.
+func IsX509Policy(ctx context.Context) bool {
+	if ctx == nil {
+		return false
+	}
+	v, _ := ctx.Value(policyConfigurationTypeContextKey{}).(policyConfigurationType)
+	return v == x509PolicyType
+}
+
+// WithSSHHostPolicy returns a context.Context with SSH host policy set.
+func WithSSHHostPolicy(ctx context.Context) context.Context {
+	return context.WithValue(ctx, policyConfigurationTypeContextKey{}, sshHostPolicyType)
+}
+
+// IsSSHHostPolicy returns if the context.Context has SSH host policy set.
+func IsSSHHostPolicy(ctx context.Context) bool {
+	if ctx == nil {
+		return false
+	}
+	v, _ := ctx.Value(policyConfigurationTypeContextKey{}).(policyConfigurationType)
+	return v == sshHostPolicyType
+}
+
+// WithSSHUserPolicy returns a context.Context with SSH user policy set.
+func WithSSHUserPolicy(ctx context.Context) context.Context {
+	return context.WithValue(ctx, policyConfigurationTypeContextKey{}, sshUserPolicyType)
+}
+
+// IsSSHUserPolicy returns if context.Context has SSH user policy set.
+func IsSSHUserPolicy(ctx context.Context) bool {
+	if ctx == nil {
+		return false
+	}
+	v, _ := ctx.Value(policyConfigurationTypeContextKey{}).(policyConfigurationType)
+	return v == sshUserPolicyType
+}
+
+type policyTypeContextKey struct{}
+
+type policyType int
+
+const (
+	allowType policyType = iota + 1
+	denyType
+)
+
+// WithAllow returns a context.Context with allow policy set.
+func WithAllow(ctx context.Context) context.Context {
+	return context.WithValue(ctx, policyTypeContextKey{}, allowType)
+}
+
+// IsAllow returns if the context.Context has allow set.
+func IsAllow(ctx context.Context) bool {
+	if ctx == nil {
+		return false
+	}
+	v, _ := ctx.Value(policyTypeContextKey{}).(policyType)
+	return v == allowType
+}
+
+// WithDeny returns a context.Context with deny set.
+func WithDeny(ctx context.Context) context.Context {
+	return context.WithValue(ctx, policyTypeContextKey{}, denyType)
+}
+
+// IsDeny returns if context.Context has deny set.
+func IsDeny(ctx context.Context) bool {
+	if ctx == nil {
+		return false
+	}
+	v, _ := ctx.Value(policyTypeContextKey{}).(policyType)
+	return v == denyType
+}
+
+func GetPrefixedCommandUsage(ctx context.Context, commandName string) string {
+	usage := "step ca policy"
```

Comment:
You should be able to get `step ca policy` using the `cli.Context` with `ctx.App.Name`, and the actual command e.g. `authority` with `ctx.Command.Name`

Reply:
Something like that was what I had in mind, but note that we have a normal `context.Context` here; not a `cli.Context`. The `ctx` does contain a `cli.Context`, but only after an action is called (when it's injected just before the command is ran), so not when the help is printed, which is when the `GetPrefixedCommandUsage` is used.

I think the pattern of using `context.Context` can (and should) be used more broadly, the main reason being making the CLI more testable, but we would need to refactor some more parts of the CLI to be able to do so. I don't think what I currently have implemented is perfect, but it gets the job done for now.